File: ios/Sources/RynBridgeCrypto/CryptoModule.swift
import Foundation
import RynBridge

public struct CryptoModule: BridgeModule, Sendable {
    public let name = "crypto"
    public let version = "0.1.0"
    public let actions: [String: ActionHandler]

    public init() {
        self.init(provider: DefaultCryptoProvider())
    }

    public init(provider: CryptoProvider) {
        actions = [
            "generateKeyPair": { _ in
                let publicKey = try await provider.generateKeyPair()
                return ["publicKey": .string(publicKey)]
            },
            "performKeyExchange": { payload in
                guard let remotePublicKey = payload["remotePublicKey"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: remotePublicKey")
                }
                let established = try await provider.performKeyExchange(remotePublicKey: remotePublicKey)
                return ["sessionEstablished": .bool(established)]
            },
            "encrypt": { payload in
                guard let data = payload["data"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: data")
                }
                let associatedData = payload["associatedData"]?.stringValue
                let result = try await provider.encrypt(data: data, associatedData: associatedData)
                return result.toPayload()
            },
            "decrypt": { payload in
                guard let ciphertext = payload["ciphertext"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: ciphertext")
                }
                guard let iv = payload["iv"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: iv")
                }
                guard let tag = payload["tag"]?.stringValue else {
                    throw RynBridgeError(code: .invalidMessage, message: "Missing required field: tag")
                }
                let associatedData = payload["associatedData"]?.stringValue
                let plaintext = try await provider.decrypt(ciphertext: ciphertext, iv: iv, tag: tag, associatedData: associatedData)
                return ["plaintext": .string(plaintext)]
            },
            "getStatus": { _ in
                let status = try await provider.getStatus()
                return status.toPayload()
            },
            "rotateKeys": { _ in
                let publicKey = try await provider.rotateKeys()
                return ["publicKey": .string(publicKey)]
            },
        ]
    }
}

File: android/crypto/src/main/kotlin/com/devryner/rynbridge/crypto/CryptoModule.kt
package com.devryner.rynbridge.crypto

import com.devryner.rynbridge.core.*

class CryptoModule(provider: CryptoProvider) : BridgeModule {
    constructor() : this(DefaultCryptoProvider())

    override val name = "crypto"
    override val version = "0.1.0"
    override val actions: Map<String, ActionHandler> = mapOf(
        "generateKeyPair" to { _ ->
            val publicKey = provider.generateKeyPair()
            mapOf("publicKey" to BridgeValue.string(publicKey))
        },
        "performKeyExchange" to { payload ->
            val remotePublicKey = payload["remotePublicKey"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: remotePublicKey")
            val established = provider.performKeyExchange(remotePublicKey)
            mapOf("sessionEstablished" to BridgeValue.bool(established))
        },
        "encrypt" to { payload ->
            val data = payload["data"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: data")
            val associatedData = payload["associatedData"]?.stringValue
            val result = provider.encrypt(data, associatedData)
            result.toPayload()
        },
        "decrypt" to { payload ->
            val ciphertext = payload["ciphertext"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: ciphertext")
            val iv = payload["iv"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: iv")
            val tag = payload["tag"]?.stringValue
                ?: throw RynBridgeError(code = ErrorCode.INVALID_MESSAGE, message = "Missing required field: tag")
            val associatedData = payload["associatedData"]?.stringValue
            val plaintext = provider.decrypt(ciphertext, iv, tag, associatedData)
            mapOf("plaintext" to BridgeValue.string(plaintext))
        },
        "getStatus" to { _ ->
            val status = provider.getStatus()
            status.toPayload()
        },
        "rotateKeys" to { _ ->
            val publicKey = provider.rotateKeys()
            mapOf("publicKey" to BridgeValue.string(publicKey))
        }
    )
}
